src/Core/Element.js

```diff
@@ -0,0 +1,267 @@
+	this.event = null
+	this.DOM = null
+	this.context = null
+	this.parent = null
+	this.host = null
+	this.next = null
+	this.prev = null
+}
+
+/**
+ * @param {*} content
+ * @return {Element}
+ */
+function elementText (content) {
+	var element = new Element(ElementText)
+
+	element.type = '#text'
+	element.children = content
+
+	return element
+}
+
+/**
+ * @param {Object} object
+ * @return {Element}
+ */
+function elementIntermediate (object) {
+	var element = new Element(ElementIntermediate)
+
+	element.context = {}
+	element.DOM = object
+
+	return element
+}
+
+/**
+ * @param {(Element|Array|List)} fragment
+ * @return {Element}
+ */
+function elementFragment (fragment) {
+	var element = new Element(ElementFragment)
+	var children = new List()
+	
+	element.type = '#fragment'
+	element.children = children
+
+	switch (fragment.constructor) {
+		case Element:
+			elementChildren(element, children, fragment, 0)
+			break
+		case Array:
+			for (var i = 0; i < fragment.length; ++i)
+				elementChildren(element, children, fragment[i], i)
+	}
+
+	return element
+}
+
+/**
+ * @param {Iterable} iterable
+ * @param {Element} element
+ */
+function elementIterable (iterable, element) {	
+	var index = 0
+
+	each(iterable, function (value) {
+		index = elementChildren(element, element.children, value, index)
+	})
+
+	return element
+}
+
+/**
+ * @param {string} summary
+ * @param {string} details
+ * @return {Element}
+ */
+function elementError (summary, details) {
+	return createElement('details', createElement('summary', summary), h('pre', details))
+}
+
+/**
+ * @param {*} child
+ * @return {Element}
+ */
+function elementUnknown (child) {
+	if (typeof child.next === 'function')
+		return elementIterable(child, elementFragment(child))
+	if (typeof child[SymbolIterator] === 'function')
+		return elementUnknown(child[SymbolIterator]())
+	else if (typeof child === 'function')
+		return elementUnknown(child())
+	else if (child instanceof Error)
+		return elementError(child+'', (child.children||'')+'\n'+child.stack)
+	else if (child instanceof Date)
+		return elementText(child)
+
+	invariant('render', 'Invalid element '+JSON.stringify(child))
+}
+
+/**
+ * @param {Element} element
+ * @param {number} signature
+ * @return {Element}
+ */
+function elementSibling (element, signature) {
+	if (signature > 0 && element.flag !== ElementPortal && isValidElement(element.children.next))
+		return element.children.next
+	else if (isValidElement(element.next))
+		return element.next
+	else
+		return elementIntermediate(DOM(null))
+}
+
+/**
+ * @param {Element} element
+ * @param {List} children
+ * @param {*} child
+ * @param {number} index
+ */
+function elementChildren (element, children, child, index) {
+	if (child == null)
+		return elementChildren(element, children, elementText(''), index)
+	else
+		switch (child.constructor) {
+			case Element:
+				if (child.key == null)
+					child.key += '#'+index
+				else if (element.keyed === false)
```

Comment:
so if the key is omitted, it's set to `#n` and element.keyed remains false?

Reply:
Yes, null+`#`+index, this is done to avoid possible problems that can arise from duplicate keys due to multiple un-keyed elements within a keyed list since we don't know when we might encounter a keyed item that will change the state of this to a keyed list.